internal/app/machined/pkg/controllers/kubeaccess/crd.go

```diff
@@ -0,0 +1,207 @@
+			Namespace: secrets.NamespaceName,
+			Type:      secrets.KubernetesType,
+			ID:        pointer.To(secrets.KubernetesID),
+			Kind:      controller.InputWeak,
+		},
+		{
+			Namespace: secrets.NamespaceName,
+			Type:      secrets.OSRootType,
+			ID:        pointer.To(secrets.OSRootID),
+			Kind:      controller.InputWeak,
+		},
+	}); err != nil {
+		return err
+	}
+
+	r.QueueReconcile()
+
+	var crdControllerCtxCancel context.CancelFunc
+
+	crdControllerErrCh := make(chan error, 1)
+
+	defer func() {
+		if crdControllerCtxCancel != nil {
+			crdControllerCtxCancel()
+
+			<-crdControllerErrCh
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil //nolint:govet
+		case <-r.EventCh():
+		case err := <-crdControllerErrCh:
+			if crdControllerCtxCancel != nil {
+				crdControllerCtxCancel()
+			}
+
+			crdControllerCtxCancel = nil
+
+			if err != nil && !errors.Is(err, context.Canceled) {
+				return fmt.Errorf("error from crd controller: %w", err)
+			}
+		}
+
+		kubeaccessConfig, err := r.Get(ctx, kubeaccess.NewConfig(config.NamespaceName, kubeaccess.ConfigID).Metadata())
+		if err != nil {
+			if !state.IsNotFoundError(err) {
+				return fmt.Errorf("error fetching kubeaccess config: %w", err)
+			}
+
+			continue
+		}
+
+		var kubeaccessConfigSpec *kubeaccess.ConfigSpec
+
+		if kubeaccessConfig != nil {
+			kubeaccessConfigSpec = kubeaccessConfig.(*kubeaccess.Config).TypedSpec()
+		}
+
+		if kubeaccessConfig == nil || kubeaccessConfigSpec == nil || !kubeaccessConfigSpec.Enabled {
+			if crdControllerCtxCancel != nil {
+				crdControllerCtxCancel()
+
+				<-crdControllerErrCh
+
+				crdControllerCtxCancel = nil
+			}
+
+			continue
+		}
+
+		kubeSecretsResources, err := r.Get(ctx, resource.NewMetadata(secrets.NamespaceName, secrets.KubernetesType, secrets.KubernetesID, resource.VersionUndefined))
+		if err != nil {
+			if !state.IsNotFoundError(err) {
+				return fmt.Errorf("error fetching kubernetes secrets: %w", err)
+			}
+
+			continue
+		}
+
+		kubeSecretsSpec := kubeSecretsResources.(*secrets.Kubernetes).TypedSpec()
+
+		osSecretsResource, err := r.Get(ctx, resource.NewMetadata(secrets.NamespaceName, secrets.OSRootType, secrets.OSRootID, resource.VersionUndefined))
+		if err != nil {
+			if !state.IsNotFoundError(err) {
+				return fmt.Errorf("error fetching os secrets: %w", err)
+			}
+
+			continue
+		}
+
+		osSecretsSpec := osSecretsResource.(*secrets.OSRoot).TypedSpec()
+
+		kubeconfig, err := clientcmd.BuildConfigFromKubeconfigGetter("", func() (*clientcmdapi.Config, error) {
+			return clientcmd.Load([]byte(kubeSecretsSpec.LocalhostAdminKubeconfig))
+		})
+		if err != nil {
+			return fmt.Errorf("error loading kubeconfig: %w", err)
+		}
+
+		if crdControllerCtxCancel != nil {
+			crdControllerCtxCancel()
+
+			<-crdControllerErrCh
+
+			crdControllerCtxCancel = nil
+		}
+
+		var crdControllerCtx context.Context
+
+		crdControllerCtx, crdControllerCtxCancel = context.WithCancel(ctx) //nolint:govet
+
+		go func() {
```

Comment:
I think you can replace it with `errgroup.Group`:
```go
eg.Go(func() error {
  err := ctrl.runCRDController(...)
  if errors.is(err, context.Canceled) {
    return nil
  }

  return
}
```

Then when stopping (can also extract to common function):

```go
stopCrdController := func() error {
  if crdControllerCtxCancel != nil {
    crdControllerCtxCancel()
    crdControllerCtxCancel = nil

    return eg.Wait()
  }
}

...
if err = stopCrdController(); err != nil {
  return err
}
```

Not a big difference, but I guess using errgroups is preferred style wise :)

Reply:
Hm, nice, this was tricky so I took the it as-is from the `discovery_service.go` but this makes sense, let me try to adapt.

Reply:
As we discussed, unfortunately this does not work since we also use the err channel in select as well. But  👍 on reusing the common cancel code.